File: ts/wasm/lib/detailed/complex_operations/role_classes/role_category_base.ts
import { IHiveChainInterface } from "../../../interfaces.js";
import { operation } from "../../../protocol.js";
import { TAccountName } from "../../hive_apps_operations/index.js";
import { IOperationSink } from "../../operation_base.js";
import { LevelBase } from "./level_base.js";

/**
 * When you create a new role category, you should extend this class.
 *
 * {@link RoleCategoryBase.authorities} should be initialized in the {@link RoleCategoryBase.init} function.
 * After initializing, Account update operation iterates over all of the levels included in the authorities object,
 * deducing the level names based on the key names.
 */
export abstract class RoleCategoryBase<AuthType extends Record<string, LevelBase<any>>> {
  public authorities!: Readonly<AuthType>;

  /**
   * Category name. It should be unique and marked `as const` for proper type narrowing
   * (See: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions)
   */
  public abstract category: string;

  /**
   * This function is responsible for gathering authority types for the given account.
   * It should return an object with keys as role names and values as authority objects.
   * Keys you return will have an impact on the role types supported by the account authority update operation
   */
  public abstract init(chain: IHiveChainInterface, account: TAccountName): Promise<void>;
  /**
   * This function should return an array of operations that will modify the user account roles.
   */
  public abstract finalize(sink: IOperationSink): operation[];

  /**
   * Indicates if any of the authority levels has changed since the last update.
   */
  public abstract get changed(): boolean;
}

File: python/wax/complex_operations/role_classes/role_category_base.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING, Generic, Iterable, TypeVar

if TYPE_CHECKING:
    from wax._private.operation_base import ConvertedToProtoOperation
    from wax.interfaces import IAuthorityDataProvider, IWaxBaseInterface
    from wax.models.basic import AccountName


AuthType = TypeVar("AuthType")
"""Type variable for the authority type."""


class RoleCategoryBase(ABC, Generic[AuthType]):
    """Base class for all role categories. Please inherit from this class when yoy create a new category."""

    @property
    @abstractmethod
    def authorities(self) -> AuthType:
        """Returns object that is filled by the `init` function."""

    @property
    @abstractmethod
    def category(self) -> str:
        """Category name. Should be unique."""

    @property
    @abstractmethod
    def changed(self) -> bool:
        """Indicates if any of the authority levels has changed since the last update."""

    @abstractmethod
    async def init(
        self,
        api: IWaxBaseInterface,
        account_name: AccountName,
        provider: IAuthorityDataProvider,
    ) -> None:
        """
        Responsible for gathering authority types for the given account.

        Args:
            api (IWaxBaseInterface | IWaxBaseInterface): instance of the wax base api subclass.
            account_name (AccountName): name of the account that will be used to gather authority data.
            provider (IAuthorityDataProvider): object that delivers authority data.
        """

    @abstractmethod
    def finalize(self, api: IWaxBaseInterface) -> Iterable[ConvertedToProtoOperation]:
        """
        Function should return iterable object of operations that will modify the user accounts roles.

        Args:
            api (IWaxBaseInterface): Instance of the wax base api subclass.
        """
